contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -634,10 +628,120 @@ limitations outlined above wil be fixed.
+can use the most appropriate method for their environment.
+ 
+Block device cleanup logic will be encapsulated in separate scripts or binaries. There will be
+several scripts that will be made available out of the box, for example:
+
+
+| Cleanup Method | Description | Suitable for Device |
+|:--------------:|-------------|:-------------------:|
+|dd-zero| Used for zeroing the device repeatedly | HDD |
+|blkdiscard| Discards sectors on the device. This cleanup method may not be supported by all devices.| SSD |
+|fs-reset| A non-secure overwrite of any existing filesystem with mkfs, followed by wipefs to remove the signature of the file system | SSD/HDD |
+|shred|Repeatedly writes random values to the block device. Less effective with wear levelling in SSDs.| HDD |
+| hdparm| Issues [ATA secure erase](https://ata.wiki.kernel.org/index.php/ATA_Secure_Erase) command to erase data on device. See ATA Secure Erase. Please note that the utility has to be supported by the device in question. | SSD/HDD |
+
+The fs-reset method is a quick and minimal approach as it does a reset of any file system, which
+works for both SSD and HDD and will be the default choice for cleaning. For SSDs, admins could
+opt for either blkdiscard which is also quite fast or hdparm. For HDDs they could opt for
+dd-zeroing or shred, which can take some time to run. Finally, the user is free to create new
+cleanup scripts of their own and have them specified in the configmap of the provisioner.
+
+The configmap from earlier section will be enhanced as follows
+```
+data:
+  storageClassMap: |
+    local-fast:
+      hostDir: "/mnt/disks"
+      mountDir: "/local-ssds"
+      blockCleanerCommand:
+         - "/scripts/dd_zero.sh"
+         - "2"
+ ```
+
+The block cleaner command will specify the script and any arguments that need to be passed to it.
+The actual block device being cleaned will be supplied to the script as an environment variable
+(LOCAL_PV_BLKDEVICE) as opposed to command line, so that the script command line has complete
+freedom on its structure. The provisioner will validate that the block device path is actually
+within the directory managed by the provisioner, to prevent destructive operations on arbitrary
+paths.
+
+The provisioner logic currently does each volume’s cleanup as a synchronous serial activity.
+However, with cleanup now potentially being a multi hour activity, the processes will have to
+be asynchronous and capable of being executed in parallel. The provisioner will ensure that all
+current asynchronous cleanup processes are tracked. Special care needs to be taken to ensure that
+when a disk has only been partially cleaned. This scenario can happen if some impatient user
+manually deletes a PV and the provisioner ends up re-creating pv ready for use (but only partially
+cleaned). This issue will be addressed in the re-design of the provisioner (details will be provided
```

Comment:
This is still in discussion, whether keeping some local state is good enough or we use a crd.

Reply:
While the redesign is still in discussion, should we just store some local state to handle this case?

Reply:
@msau42 I think there are couple of other options that might be worth considering
1. Since block is still not enabled in provisioner and API, and the concerns about impatient admin only seem relevant there, we could go with the current code without any signifcant risks.
2. We could actually go with the CRD approach for book keeping here as it seems the least controversial and is likely to be used even in the new design.

Reply:
Ok, let's leave the design open, but let's check in the code (disabled by default).